source/common/protobuf/utility.h

```diff
@@ -215,6 +215,14 @@ class MessageUtil {
     return HashUtil::xxHash64(text);
   }
 
+  /**
+   * A hash function that stable hashes known Any in the message. This is much slower than
```

Comment:
This seems incredibly fragile. Shouldn't the above function handle this correctly? Is there a protobuf issue tracking this that we can link to / TODO?

Reply:
Yeah, this seems pretty bad due to the expectation that debug string is stable in any way. I feel it might be one of the least bad things for now though. We might want to move away from using any form of protobuf hashing in the future, instead opting to only consider explicit UUIDs or versions.

Reply:
OK I updated the implementation to use TextFormat directly with more settings to TextFormat::Printer explicitly, as well as more test to validate map as well. Does this look better now?

Reply:
Sure, but TBH, I still think we should take a long term (beyond this PR) think and see if we can distance ourselves from proto hashing. There have been too many bugs here and proto is unlikely to standardize a hash due to the implications on serialization.

Reply:
+1, I'm not sure how the developer is supposed to know which of these to use. We need a holistic solution that avoids this problem entirely.

Reply:
We can use this function as the only proto hashing, the only downside is this is about 10-20x slower than current `MessageUtil::hash` but if we agree this is a necessary performance tax to pay then we can do that. 10-20x slower is not too bad and the serialization is still way faster than jsonConvert, which we are using a lot. The perf saving by deduplicating is larger. WDYT?

Reply:
What about just stepping back and getting rid of hashes entirely and using the message differencer? Presumably that works for any? At this I suspect we keep the raw protos around in almost all cases for config_dump so there really aren't any memory savings, and I wonder how much slower the differencer would really be as the hash has to iterate through the entire message anyway.

Reply:
The hash is needed primarily for map keys, not for saving memory. Message differences cannot be used in that case.

Reply:
In which case is it used that way? Most cases that I know of actually hash on name, and then use the hash to check for differences.

Reply:
@mattklein123 
FCM: https://github.com/envoyproxy/envoy/blob/master/source/server/filter_chain_manager_impl.cc#L33 (MessageUtil implements std::hash and std::equals_to)
ConfigProvider: https://github.com/envoyproxy/envoy/blob/master/source/common/config/config_provider_impl.h#L414
RDS: https://github.com/envoyproxy/envoy/blob/master/source/common/router/rds_impl.cc#L230
and SDS (in this PR).

Those are to deduplicate ConfigSource (which can come with gRPC credentials extensions in Any)

Reply:
OK. Should the long term plan be that each resource has a unique name and that we use message differencer? I think this is what @htuch suggested? It would be nice to agree on the long term plan and put in relevant TODOs and possibly fix this in v3?

Reply:
(sorry to interrupt) I think using the resource version should take preference over hashing. There may be cases when the control plan wants to trigger a reload for a config that is structurally the same (e.g. references some modified file or resource by name).

Reply:
@mattklein123 @kyessenov Both SGTM, filed https://github.com/envoyproxy/envoy/issues/8301 for further discussion, it is beyond the scope of this PR.

@mattklein123 how do you want this PR land? I can make this as default hash and use everywhere (solves potential issue in other places like RDS, but slightly less performant), or leave as is to at least solve SDS issue.

Reply:
My personal feeling is that we should use a single hash that works in all cases, even if it performs worse, until we see perf traces that prove that we shouldn't do that. @htuch any opinion?

Reply:
Yep, single hash function sounds best. If we can simplify and then only add in the necessary complexity for performance reasons, that would be ideal.

Reply:
Done and updated title/description.